src/tokenizer.1
.TH TOKENIZER 1 2023-07-14
.SH NAME
\fBtokenizer\fR \(en convert source code into integer vectors
.SH SYNOPSIS
\fBtokenizer\fR [\fB\-acgs\fR | \fB-B\fR | \fB-b\fP] [\fB\-fLV\fP] [\fB\-i \fIfile\fR] [\fB\-l \fIlang\fR] [\fB\-o \fIopt\fR] [\fB\-t \fIsep\fR] [\fIfile ...\fR]
.SH DESCRIPTION
The \fBtokenizer\fR utility converts source code specified as files in
its command line or provided through its standard input into one of several
supported output formats.

By default the \fItokenizer\fR outputs numerical token codes.
The correspondence between tokens and values is not a stable API,
but it can always be established by running \fItokenizer\fP \fb-L\fP.

.RS 3
.IP \(bu
Characters are output as values in the range 0\(en255.
.IP \(bu
The language's keywords, other multi-character tokens (e.g. \fC+=\fP),
and numbers are output as values in the ranges that can be seen by running
\fItokenizer\fP \fb-L\fP.
The numbers for the same keywords are the same between different languages,
allowing the tokenizer to be used to train models through transfer learning.
.IP \(bu
Numbers are output on a base-ten logarithmic scale centered around zero
and extending by less than 400 values around it.
.IP \(bu
Other constants (e.g. characters and strings) are represented as an integer
corresponding to that token type.
.IP \(bu
Comments are mapped into a single integer.
.IP \(bu
White space is not included in the output, unless the \fB-a\fP
option is specified.
.IP \(bu
Identifiers are output as a value starting at \fCFIRST_IDENTIFIER\fP.
Identifiers with the same name within a scope get the same value.
Identifiers in outer scopes are visible with the same value in inner scopes.
Identfiers in first seen in a scope are not visible outside that scope.
.IP \(bu
When the \fB-a\fP option is specified, comments and strings have their
token value followed by a value of their content's hash
as a value starting at \fCFIRST_HASH_CONTENT\fP.
Moreover horizontal space (spaces and tabs) is output, run length
encoded as seen in the output of the \fB-L\fP option.
.RE

The default form is suitable for training machine-learning models.
With global scoping, it is also well-suited for Type-1 (exact in code)
clone detection.

.SH OPTIONS
The behavior of the \fItokenizer\fR utility can be controlled
through the following command-line options.

.RS 3

.TP
.B "-a"
Output distinct token values for all input, including horizontal white space
and the content of strings and comments.
This option can be used for Type-0 (exact in all text) clone detection.

.TP
.B -B
Output the type of each token (ID, KW, NUM, or TOK) and the actual token,
one entry per line.
Comments and strings are output with their content replaced by an
ellipsis (...).
All other tokens are output in their original form.

.TP
.B -b
Similar to the \fB-B\fP output format, but without the token type
prefix.
This output format is suitable for passing to line-difference programs
in order to process the code at the level of individual tokens.

.TP
.B -c
Compress token values so that all identifiers, numbers, and basic types
are treated as the same.
In the numeric output form
all numbers are represented with the value corresponding to the token symbol
\fCANY_NUMBER\fP,
all basic types and their modifiers (and sequences of them)
are represented with the value corresponding to the token symbol
\fCANY_TYPE\fP,
all identifiers with the value corresponding to the token symbol
\fCANY_IDENTIFIER\fP.
This option can be used for Type-2 (near or renamed) clone detection.

.TP
.B -f
Identify each read file, before outputting its tokens.
The file name is output on a separate line prefixed with the letter "F".
This is the same convention as that expected by the clone detector
.IR mpcd (1).

.TP
.B -g
Global scoping.
All identifiers with the same name are allocated to the same
token, irrespective of their scope.
This option is useful for Type-1 (exact in code) clone detection.

.TP
.BI "-i " file
Process the files listed in the specified file.
If the file name is "\(en", then the list of files to process
is read from the program's standard input.

.TP
.B -L
List the values and corresponding strings associated with the
\fItokenizer\fP tokens, and then exit.

.TP
.BI "-l " lang
Specify the input language.
Currently the following languages are supported:
\fIC\fP,
\fICSharp\fP (or \fIC#\fP),
\fIC++\fP,
\fIGo\fP,
\fIJava\fP,
\fIJavaScript\fP,
\fIPHP\fP,
\fIPython\fP,
\fIRust\fP,
\fITypeScript\fP.

.TP
.BI "-o " opt
Specify a language-specific processing option.
Multiple \fB-o\fP arguments may be provided.
The following processing options are supported.

.RS 3

.TP
.B file
Output a single vector for the whole file.

.TP
.B line
Output one vector per input line.
Input and output lines are synchronized, allowing the tracking
of tokens to their original line position.

.TP
.B method
Output a vector for the contents of each method.

.TP
.B statement
Output a vector for each statement.
.LP
.RE

.TP
.B -s
Output symbolic token values.
.RS 3
.IP \(bu
Single character tokens and reserved words are output in their original form.
.IP \(bu
Other language tokens are output as a symbol representing them.
.IP \(bu
Numbers are output in scientific notation as 1, prefixed by "~",
and followed by a varying exponent according to the number's base-ten logarithm.
.IP \(bu
Identifiers are prefixed by "ID:" and followed by an integer corresponding to
the identifier number defined for numeric token values.
.IP \(bu
When the \fB-a\fP option is specified, comments, strings, and single-quote
literals have their
token followed by a value of their content's hash prefixed by "HASH:".
.RE

.TP
.BI "-t " sep
Specify the separator character for non-line output.
By default this is tab for numeric tokens and space for symbolic output.

.TP
.B "-V"
Display the program's version number and exit.

.RE

.SH EXAMPLES
.PP
Process a C file to obtain symbolic output.
.ft C
.nf
tokenizer -l C -s file.c
.ft P
.fi

.PP
Process a Java file obtaining a numeric vector per method.
.ft C
.nf
tokenizer -l Java -o method File.java
.ft P
.fi

.PP
Produce a token-by-token difference between the current version of the
file \fCtokenizer.cpp\fP and the one in version v1.1.
.ft C
.nf
diff <(git show v1.1:./tokenizer.cpp | tokenizer -l C++ -b) \\
  <(tokenizer -l C++ -b tokenizer.cpp)
.ft P
.fi

.PP
List Type-2 (near or renamed) clones in the \fItokenizer\fP source code.
.ft C
.nf
tokenizer -l C++ -c -f -o line *.cpp *.h | mpcd
.ft P
.fi

.SH DIAGNOSTICS
An error is displayed when an end of file is encountered while processing
a block comment or a character or string literal.

.SH SEE ALSO
.IR mpcd (1)
\(em modular performant clone detector.

.SH AUTHORS
Written by Diomidis Spinellis.

.SH BUGS
The tokenizer is more permissive than the corresponding language specifications.
For example, number literals can contain arbitrary letters in them,
in addition to the recognized \fC0x\fP and \fC0b\fP prefixes.
Also string and character literals can contain arbitrary backslash-escape
sequences, in addition the defined ones.
.PP
Continuation lines are not supported.
.PP
Non-ASCII identifiers are not supported.
.PP
The processing context may get confused by unbalanced braces occurring in
C preprocessor macro definitions.
.PP
The Python tokenizer does not support processing options and identifier
scoping.
.PP
The Rust tokenizer handles tuple indices and literal suffixes as separate
lexical elements (numbers and identifiers, respectively).


src/tokenizer.cpp
/*-
 * Copyright 2014-2023 Diomidis Spinellis
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

#include <cstring>
#include <string>
#include <fstream>
#include <iostream>
#include <ostream>
#include <optional>
#include <vector>

#include "errno.h"
#include "unistd.h"

#include "SymbolTable.h"
#include "CTokenizer.h"
#include "CppTokenizer.h"
#include "CSharpTokenizer.h"
#include "GoTokenizer.h"
#include "JavaTokenizer.h"
#include "JavaScriptTokenizer.h"
#include "PHPTokenizer.h"
#include "PythonTokenizer.h"
#include "RustTokenizer.h"

const char version[] = "2.8.1";

// Command-line option values
static bool all_contents = false;
static bool symbolic_output = false;
static bool compress_ids = false;
static bool show_file_name = false;
static enum output_type {
	ot_tokens,	// Numeric or symbolic tokens
	ot_break, 	// Original tokens broken into lines
	ot_type_break,	// As above, tokens preceded by their type
} output_type = ot_tokens;
static std::string lang("Java");
static std::vector<std::string> processing_opt;
static char separator;

/*
 * Process and print the metrics of the specified stream,
 * which is identified with the specified filename.
 */
static void
process_file(std::istream &in, std::string filename)
{
	CharSource cs(in);
	TokenizerBase *t;

	if (lang == "C")
		t = new CTokenizer(cs, filename, processing_opt);
	else if (lang == "CSharp" || lang == "C#")
		t = new CSharpTokenizer(cs, filename, processing_opt);
	else if (lang == "Go")
		t = new GoTokenizer(cs, filename, processing_opt);
	else if (lang == "Java")
		t = new JavaTokenizer(cs, filename, processing_opt);
	else if (lang == "JavaScript")
		t = new JavaScriptTokenizer(cs, filename, processing_opt);
	else if (lang == "C++")
		t = new CppTokenizer(cs, filename, processing_opt);
	else if (lang == "PHP")
		t = new PHPTokenizer(cs, filename, processing_opt);
	else if (lang == "Python")
		t = new PythonTokenizer(cs, filename, processing_opt);
	else if (lang == "Rust")
		t = new RustTokenizer(cs, filename, processing_opt);
	else if (lang == "TypeScript")
		t = new JavaScriptTokenizer(cs, filename, processing_opt,
				Keyword::L_TypeScript);
	else {
		std::cerr << "Unknown language specified." << std::endl;
		std::cerr << "The following languages are supported:" << std::endl;
		std::cerr << "\tC" << std::endl;
		std::cerr << "\tCSharp (or C#)" << std::endl;
		std::cerr << "\tC++" << std::endl;
		std::cerr << "\tGo" << std::endl;
		std::cerr << "\tJava" << std::endl;
		std::cerr << "\tJavaScript" << std::endl;
		std::cerr << "\tPHP" << std::endl;
		std::cerr << "\tPython" << std::endl;
		std::cerr << "\tRust" << std::endl;
		std::cerr << "\tTypeScript" << std::endl;
		exit(EXIT_FAILURE);
	}

	t->set_separator(separator ? separator : ' ');
	t->set_all_contents(all_contents);
	switch (output_type) {
	case ot_tokens:
		if (symbolic_output) {
			if (compress_ids)
				t->type_tokenize();
			else
				t->symbolic_tokenize();
		} else {
			t->set_separator(separator ? separator : '\t');
			t->numeric_tokenize(compress_ids);
		}
		break;
	case ot_type_break:
		t->type_code_tokenize();
		break;
	case ot_break:
		t->code_tokenize();
		break;
	}
}

// List the values of all tokens
static void
list_tokens()
{
	// Characters (ASCII assumed)
	for (char c = ' '; c <= '~'; c++)
		std::cout << static_cast<int>(c) << '\t' << c << std::endl;

	// Character symbol tokens (e.g. +=)
	Token t;
	for (auto ti: t.token_symbol_view())
		std::cout << ti.first << '\t' << ti.second << std::endl;

	// Keywords (e.g. "if")
	// Language is irrelevant for token_keyword_view
	Keyword k(Keyword::L_C);
	for (auto ki: k.token_keyword_view())
		std::cout << ki.first << '\t' << ki.second << std::endl;

	// Compressed token identifiers
	std::cout << TokenId::ANY_TYPE << "\tANY_TYPE" << std::endl;
	std::cout << TokenId::ANY_HASH << "\tANY_HASH" << std::endl;
	std::cout << TokenId::ANY_NUMBER << "\tANY_NUMBER" << std::endl;
	std::cout << TokenId::ANY_IDENTIFIER << "\tANY_IDENTIFIER" << std::endl;

	// Numbers
	for (long double d = 1e-308L; d < 1; d *= 10)
		std::cout << TokenizerBase::compress(static_cast<double>(d)) << '\t' << d << std::endl;
	std::cout << TokenId::NUMBER_ZERO << "\t0" << std::endl;
	for (long double d = 1; d < 1e309L; d *= 10)
		std::cout << TokenizerBase::compress(static_cast<double>(d)) << '\t' << d << std::endl;

	// RLE horizontal space
	for (int i = 2; i <= TokenId::RLE_MAX; ++i)
		std::cout << TokenId::RLE_SPACE + i << '\t' << "' ' * " << i << std::endl;
	for (int i = 2; i <= TokenId::RLE_MAX; ++i)
		std::cout << TokenId::RLE_TAB + i << '\t' << "'\\t' * " << i << std::endl;

	// Other token values
	std::cout << TokenId::NUMBER_INFINITE << "\tINFINITE" << std::endl;
	std::cout << TokenId::NUMBER_NAN << "\tNAN" << std::endl;
	std::cout << TokenId::FIRST_IDENTIFIER << "\tFIRST_IDENTIFIER" << std::endl;
	std::cout << TokenId::HASHED_CONTENT << "\tFIST_HASHED_CONTENT" << std::endl;
}

// Open and process the specified file
static void
process_named_file(std::string filename)
{
	std::ifstream in;

	in.open(filename, std::ifstream::in);
	if (!in.good()) {
		std::cerr << "Unable to open " << filename <<
			": " << strerror(errno) << std::endl;
		exit(EXIT_FAILURE);
	}
	if (show_file_name)
		std::cout << "F" << filename << std::endl;
	process_file(in, filename);
	in.close();
}

// Process the files listed in the specified input stream
static void
process_files_from_stream(std::istream &in)
{
	std::string filename;

	while (std::getline(in, filename))
		process_named_file(filename);
}

/* Read code and output its constituent tokens */
int
main(int argc, char * const argv[])
{
	int opt;
	std::optional<std::string> files_list(std::nullopt);

	while ((opt = getopt(argc, argv, "aBbcfgi:Ll:o:st:V")) != -1)
		switch (opt) {
		case 'a':
			all_contents = true;
			break;
		case 'B':
			output_type = ot_type_break;
			break;
		case 'b':
			output_type = ot_break;
			break;
		case 'c':
			compress_ids = true;
			break;
		case 'f':
			show_file_name = true;
			break;
		case 'g':
			SymbolTable::disable_scoping();
			break;
		case 'i':
			files_list = optarg;
			break;
		case 'L':
			list_tokens();
			exit(EXIT_SUCCESS);
			break;
		case 'l':
			lang = optarg;
			break;
		case 'o':
			processing_opt.push_back(optarg);
			break;
		case 's':
			symbolic_output = true;
			break;
		case 't':
			separator = *optarg;
			break;
		case 'V':
			std::cout << "tokenizer " << version << std::endl;
			exit(EXIT_SUCCESS);
		default: /* ? */
			std::cerr << "Usage: " << argv[0] <<
				"  [-acgs | -B | -b] [-fV] [-i file] [-l lang] [-o opt] [-t sep] [file ...]" << std::endl;
			exit(EXIT_FAILURE);
		}

	if (argv[optind] && files_list.has_value()) {
		std::cerr << "Specify either an input file list or"
			" command-line arguments; not both." << std::endl;
		exit(EXIT_FAILURE);
	}

	// Process tokens from standard input
	if (!argv[optind] && !files_list.has_value()) {
		process_file(std::cin, "-");
		exit(EXIT_SUCCESS);
	}

	if (files_list.has_value()) {

		if (files_list.value() == "-") {
			process_files_from_stream(std::cin);
		} else {
			std::ifstream file_list_stream(files_list.value());
			if (!file_list_stream) {
				std::cerr << "Unable to open "
					<< files_list.value() <<
					": " << strerror(errno) << std::endl;
				exit(EXIT_FAILURE);
			}
			process_files_from_stream(file_list_stream);
		}

	}

	// Read from files specified as arguments
	for (; argv[optind]; optind++)
		process_named_file(argv[optind]);

	exit(EXIT_SUCCESS);
}
